File: exercises/cpp/08_leading_trailing/leading_trailing.cpp
/*
 * ==============================================================================
 * Exercise 8: LEADING and TRAILING Sets for Operator Grammars
 * ==============================================================================
 *
 * CONCEPT:
 * In operator precedence parsing, LEADING and TRAILING sets determine the
 * precedence relations between terminals. An operator grammar has no two
 * adjacent non-terminals in any production right-hand side.
 *
 * LEADING(A) is the set of terminals that can appear as the first terminal
 * in a string derived from non-terminal A (possibly preceded by non-terminals).
 * TRAILING(A) is the set of terminals that can appear as the last terminal
 * in a string derived from A (possibly followed by non-terminals).
 *
 * ALGORITHM:
 * Iterative fixed-point computation:
 *   1. For each production A -> X1 X2 ..., scan from the left. Add the first
 *      terminal found to LEADING(A). If a non-terminal B appears before that
 *      terminal, mark that LEADING(B) should propagate to LEADING(A).
 *   2. Repeat propagation until no set changes (fixed-point).
 *   3. Analogous process scanning from the right for TRAILING.
 *
 * TIME COMPLEXITY:  O(|P| * |V|) per iteration; converges in O(|N|) iterations.
 * SPACE COMPLEXITY: O(|N| * |T|) for the sets.
 *
 * EXAMPLE INPUT:
 *   E -> E + T | T
 *   T -> T * F | F
 *   F -> ( E ) | id
 *
 * EXAMPLE OUTPUT:
 *   LEADING(E) = { (, *, +, id }
 *   TRAILING(E) = { ), *, +, id }
 * ==============================================================================
 */

#include <iostream>
#include <string>
#include <vector>
#include <set>
#include <map>
#include <sstream>
#include <algorithm>

/* A production: lhs -> rhs_symbols */
struct Production {
    std::string lhs;
    std::vector<std::string> rhs;
};

class LeadingTrailingCalculator {
private:
    std::vector<Production> productions_;
    std::set<std::string> non_terminals_;
    std::set<std::string> terminals_;
    std::vector<std::string> nt_order_; /* Insertion-ordered non-terminals */

    std::map<std::string, std::set<std::string>> leading_;
    std::map<std::string, std::set<std::string>> trailing_;

    bool is_non_terminal(const std::string &sym) const {
        return non_terminals_.count(sym) > 0;
    }

public:
    void add_non_terminal(const std::string &nt) {
        if (non_terminals_.insert(nt).second) {
            nt_order_.push_back(nt);
        }
    }

    void add_production(const std::string &lhs,
                        const std::vector<std::string> &rhs) {
        add_non_terminal(lhs);
        productions_.push_back({lhs, rhs});
    }

    void identify_terminals() {
        for (const auto &prod : productions_) {
            for (const auto &sym : prod.rhs) {
                if (!is_non_terminal(sym)) {
                    terminals_.insert(sym);
                }
            }
        }
    }

    void compute_leading() {
        /* Initialize empty sets */
        for (const auto &nt : non_terminals_) {
            leading_[nt] = {};
        }

        /* Initial pass: direct terminals from left scan */
        for (const auto &prod : productions_) {
            for (const auto &sym : prod.rhs) {
                if (terminals_.count(sym)) {
                    leading_[prod.lhs].insert(sym);
                    break;
                } else if (is_non_terminal(sym)) {
                    continue; /* Skip leading non-terminals */
                } else {
                    break;
                }
            }
        }

        /* Fixed-point propagation */
        bool changed = true;
        while (changed) {
            changed = false;
            for (const auto &prod : productions_) {
                for (const auto &sym : prod.rhs) {
                    if (is_non_terminal(sym)) {
                        size_t before = leading_[prod.lhs].size();
                        leading_[prod.lhs].insert(
                            leading_[sym].begin(), leading_[sym].end());
                        if (leading_[prod.lhs].size() > before) {
                            changed = true;
                        }
                    } else {
                        break; /* Terminal reached */
                    }
                }
            }
        }
    }

    void compute_trailing() {
        for (const auto &nt : non_terminals_) {
            trailing_[nt] = {};
        }

        /* Initial pass: direct terminals from right scan */
        for (const auto &prod : productions_) {
            for (int i = static_cast<int>(prod.rhs.size()) - 1; i >= 0; i--) {
                const std::string &sym = prod.rhs[i];
                if (terminals_.count(sym)) {
                    trailing_[prod.lhs].insert(sym);
                    break;
                } else if (is_non_terminal(sym)) {
                    continue;
                } else {
                    break;
                }
            }
        }

        /* Fixed-point propagation */
        bool changed = true;
        while (changed) {
            changed = false;
            for (const auto &prod : productions_) {
                for (int i = static_cast<int>(prod.rhs.size()) - 1; i >= 0; i--) {
                    const std::string &sym = prod.rhs[i];
                    if (is_non_terminal(sym)) {
                        size_t before = trailing_[prod.lhs].size();
                        trailing_[prod.lhs].insert(
                            trailing_[sym].begin(), trailing_[sym].end());
                        if (trailing_[prod.lhs].size() > before) {
                            changed = true;
                        }
                    } else {
                        break;
                    }
                }
            }
        }
    }

    void print_grammar() const {
        std::cout << "\n==================================================\n";
        std::cout << "  Grammar Productions\n";
        std::cout << "==================================================\n";

        /* Group productions by LHS for pretty printing */
        for (const auto &nt : nt_order_) {
            bool first_alt = true;
            std::cout << "  " << nt << " -> ";
            for (const auto &prod : productions_) {
                if (prod.lhs != nt) continue;
                if (!first_alt) std::cout << " | ";
                first_alt = false;
                for (size_t i = 0; i < prod.rhs.size(); i++) {
                    if (i > 0) std::cout << " ";
                    std::cout << prod.rhs[i];
                }
            }
            std::cout << "\n";
        }
        std::cout << "==================================================\n";
    }

    void print_symbols() const {
        std::cout << "\n  Non-terminals: { ";
        bool first = true;
        for (const auto &nt : nt_order_) {
            if (!first) std::cout << ", ";
            std::cout << nt;
            first = false;
        }
        std::cout << " }\n";

        std::cout << "  Terminals:     { ";
        first = true;
        for (const auto &t : terminals_) {
            if (!first) std::cout << ", ";
            std::cout << t;
            first = false;
        }
        std::cout << " }\n";
    }

    void print_sets(const std::string &name,
                    const std::map<std::string, std::set<std::string>> &sets) const {
        std::cout << "\n==================================================\n";
        std::cout << "  " << name << " Sets\n";
        std::cout << "==================================================\n";

        for (const auto &nt : nt_order_) {
            auto it = sets.find(nt);
            if (it == sets.end()) continue;

            std::cout << "  " << name << "(" << nt << ") = { ";
            bool first = true;
            for (const auto &t : it->second) {
                if (!first) std::cout << ", ";
                std::cout << t;
                first = false;
            }
            std::cout << " }\n";
        }
        std::cout << "==================================================\n";
    }

    void print_results() const {
        print_sets("LEADING", leading_);
        print_sets("TRAILING", trailing_);
    }
};

int main() {
    std::cout << "==================================================\n";
    std::cout << "  LEADING and TRAILING Sets Calculator\n";
    std::cout << "  (Operator Precedence Parsing)\n";
    std::cout << "==================================================\n";

    LeadingTrailingCalculator calc;

    /* Register non-terminals first to control ordering */
    calc.add_non_terminal("E");
    calc.add_non_terminal("T");
    calc.add_non_terminal("F");

    /* Grammar: E -> E+T | T, T -> T*F | F, F -> (E) | id */
    calc.add_production("E", {"E", "+", "T"});
    calc.add_production("E", {"T"});
    calc.add_production("T", {"T", "*", "F"});
    calc.add_production("T", {"F"});
    calc.add_production("F", {"(", "E", ")"});
    calc.add_production("F", {"id"});

    calc.identify_terminals();
    calc.print_grammar();
    calc.print_symbols();

    calc.compute_leading();
    calc.compute_trailing();
    calc.print_results();

    /* Second example */
    std::cout << "\n\n##################################################\n";
    std::cout << "  Second Example Grammar\n";
    std::cout << "##################################################\n";

    LeadingTrailingCalculator calc2;
    calc2.add_non_terminal("S");
    calc2.add_non_terminal("A");
    calc2.add_non_terminal("B");

    calc2.add_production("S", {"S", "+", "A"});
    calc2.add_production("S", {"A"});
    calc2.add_production("A", {"A", "*", "B"});
    calc2.add_production("A", {"B"});
    calc2.add_production("B", {"(", "S", ")"});
    calc2.add_production("B", {"a"});

    calc2.identify_terminals();
    calc2.print_grammar();
    calc2.print_symbols();

    calc2.compute_leading();
    calc2.compute_trailing();
    calc2.print_results();

    return 0;
}

File: exercises/python/08_leading_trailing/leading_trailing.py
"""
==============================================================================
Exercise 8: LEADING and TRAILING Sets for Operator Grammars
==============================================================================

CONCEPT:
In operator precedence parsing, LEADING and TRAILING sets are used to
establish precedence relations between terminal symbols. An operator grammar
is a context-free grammar where no production right-hand side contains two
adjacent non-terminals.

LEADING(A) is the set of terminals 'a' such that there exists a derivation
A =>* Ba... (i.e., 'a' can appear as the first terminal in some string
derived from non-terminal A, possibly preceded by non-terminals).

TRAILING(A) is the set of terminals 'a' such that there exists a derivation
A =>* ...aB (i.e., 'a' can appear as the last terminal in some string
derived from non-terminal A, possibly followed by non-terminals).

ALGORITHM:
We use an iterative fixed-point computation:
  1. Initialize LEADING(A) with terminals that appear as the first terminal
     in any production of A (scanning from the left).
  2. If a production A -> B... starts with a non-terminal B, add
     LEADING(B) to LEADING(A).
  3. Repeat step 2 until no changes occur (fixed-point).
  Similarly for TRAILING but scanning from the right.

TIME COMPLEXITY:  O(|P| * |V|) where |P| is total production length and
                  |V| is the number of grammar symbols.
SPACE COMPLEXITY: O(|N| * |T|) for storing the sets, where |N| is the
                  number of non-terminals and |T| the number of terminals.

EXAMPLE INPUT:
  E -> E + T | T
  T -> T * F | F
  F -> ( E ) | id

EXAMPLE OUTPUT:
  LEADING(E) = { +, *, (, id }
  LEADING(T) = { *, (, id }
  LEADING(F) = { (, id }
  TRAILING(E) = { +, *, ), id }
  TRAILING(T) = { *, ), id }
  TRAILING(F) = { ), id }
==============================================================================
"""


def parse_grammar(grammar_text):
    """
    Parse a grammar specification into a dictionary.
    Each key is a non-terminal, and the value is a list of productions,
    where each production is a list of symbols.
    """
    grammar = {}
    non_terminals = set()
    terminals = set()

    for line in grammar_text.strip().split("\n"):
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        lhs, rhs = line.split("->")
        lhs = lhs.strip()
        non_terminals.add(lhs)

        alternatives = rhs.split("|")
        productions = []
        for alt in alternatives:
            symbols = alt.strip().split()
            productions.append(symbols)

        if lhs in grammar:
            grammar[lhs].extend(productions)
        else:
            grammar[lhs] = productions

    # Identify terminals: any symbol that is not a non-terminal
    for lhs, productions in grammar.items():
        for production in productions:
            for symbol in production:
                if symbol not in non_terminals:
                    terminals.add(symbol)

    return grammar, non_terminals, terminals


def compute_leading(grammar, non_terminals, terminals):
    """
    Compute LEADING sets using an iterative fixed-point algorithm.

    Rule 1: If A -> a... or A -> Ba..., then a is in LEADING(A).
    Rule 2: If A -> B..., then LEADING(B) is a subset of LEADING(A).
    Repeat until stable.
    """
    leading = {nt: set() for nt in non_terminals}

    # Initial pass: direct terminals from productions
    for non_terminal, productions in grammar.items():
        for production in productions:
            for symbol in production:
                if symbol in terminals:
                    leading[non_terminal].add(symbol)
                    break  # Only the first terminal matters
                elif symbol in non_terminals:
                    # Non-terminal at the start; we will propagate later.
                    # But keep scanning in case a terminal follows immediately.
                    continue
                break  # Should not reach here with well-formed grammar

    # Iterative fixed-point: propagate through non-terminal prefixes
    changed = True
    while changed:
        changed = False
        for non_terminal, productions in grammar.items():
            for production in productions:
                for symbol in production:
                    if symbol in non_terminals:
                        # Add LEADING of this non-terminal
                        before_size = len(leading[non_terminal])
                        leading[non_terminal] |= leading[symbol]
                        if len(leading[non_terminal]) > before_size:
                            changed = True
                        # Continue scanning; the non-terminal might be
                        # followed by a terminal that is also a leading
                        # terminal (but actually, the first terminal we
                        # encounter from the left is the only one to add
                        # directly, and the non-terminal prefix means
                        # we propagate). We break after processing the
                        # prefix of non-terminals and the first terminal.
                        continue
                    elif symbol in terminals:
                        # Already added in initial pass
                        break
                    break

    return leading


def compute_trailing(grammar, non_terminals, terminals):
    """
    Compute TRAILING sets using an iterative fixed-point algorithm.

    Rule 1: If A -> ...a or A -> ...aB, then a is in TRAILING(A).
    Rule 2: If A -> ...B, then TRAILING(B) is a subset of TRAILING(A).
    Repeat until stable.
    """
    trailing = {nt: set() for nt in non_terminals}

    # Initial pass: scan from the right of each production
    for non_terminal, productions in grammar.items():
        for production in productions:
            for symbol in reversed(production):
                if symbol in terminals:
                    trailing[non_terminal].add(symbol)
                    break
                elif symbol in non_terminals:
                    continue
                break

    # Iterative fixed-point: propagate through non-terminal suffixes
    changed = True
    while changed:
        changed = False
        for non_terminal, productions in grammar.items():
            for production in productions:
                for symbol in reversed(production):
                    if symbol in non_terminals:
                        before_size = len(trailing[non_terminal])
                        trailing[non_terminal] |= trailing[symbol]
                        if len(trailing[non_terminal]) > before_size:
                            changed = True
                        continue
                    elif symbol in terminals:
                        break
                    break

    return trailing


def print_sets(title, sets, non_terminal_order):
    """Print LEADING or TRAILING sets in a formatted table."""
    print(f"\n{'=' * 50}")
    print(f"  {title}")
    print(f"{'=' * 50}")
    for nt in non_terminal_order:
        if nt in sets:
            elements = ", ".join(sorted(sets[nt]))
            print(f"  {title.split()[0]}({nt}) = {{ {elements} }}")
    print(f"{'=' * 50}")


def print_grammar(grammar, non_terminal_order):
    """Print the grammar in a readable format."""
    print(f"\n{'=' * 50}")
    print("  Grammar Productions")
    print(f"{'=' * 50}")
    for nt in non_terminal_order:
        if nt in grammar:
            alternatives = [" ".join(prod) for prod in grammar[nt]]
            print(f"  {nt} -> {' | '.join(alternatives)}")
    print(f"{'=' * 50}")


def main():
    # Example grammar: E -> E+T | T, T -> T*F | F, F -> (E) | id
    grammar_text = """
        E -> E + T | T
        T -> T * F | F
        F -> ( E ) | id
    """

    print("=" * 50)
    print("  LEADING and TRAILING Sets Calculator")
    print("  (Operator Precedence Parsing)")
    print("=" * 50)

    grammar, non_terminals, terminals = parse_grammar(grammar_text)

    # Maintain a consistent ordering for output
    non_terminal_order = ["E", "T", "F"]
    # Fall back to sorted order if grammar has different non-terminals
    non_terminal_order = [nt for nt in non_terminal_order if nt in non_terminals]
    for nt in sorted(non_terminals):
        if nt not in non_terminal_order:
            non_terminal_order.append(nt)

    print_grammar(grammar, non_terminal_order)

    print(f"\n  Non-terminals: {{ {', '.join(sorted(non_terminals))} }}")
    print(f"  Terminals:     {{ {', '.join(sorted(terminals))} }}")

    leading = compute_leading(grammar, non_terminals, terminals)
    trailing = compute_trailing(grammar, non_terminals, terminals)

    print_sets("LEADING Sets", leading, non_terminal_order)
    print_sets("TRAILING Sets", trailing, non_terminal_order)

    # Demonstrate with a second grammar
    print("\n\n" + "#" * 50)
    print("  Second Example Grammar")
    print("#" * 50)

    grammar_text_2 = """
        S -> S + A | A
        A -> A * B | B
        B -> ( S ) | a
    """

    grammar2, non_terminals2, terminals2 = parse_grammar(grammar_text_2)
    nt_order2 = ["S", "A", "B"]
    nt_order2 = [nt for nt in nt_order2 if nt in non_terminals2]

    print_grammar(grammar2, nt_order2)

    leading2 = compute_leading(grammar2, non_terminals2, terminals2)
    trailing2 = compute_trailing(grammar2, non_terminals2, terminals2)

    print_sets("LEADING Sets", leading2, nt_order2)
    print_sets("TRAILING Sets", trailing2, nt_order2)


if __name__ == "__main__":
    main()
